cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/ZooKeeperRoller.java

```diff
@@ -53,60 +57,61 @@ public Future<Void> maybeRollingUpdate(Reconciliation reconciliation, Labels sel
                         } else {
-                            LOGGER.debugCr(reconciliation, "Pod {} does not need to be rolled", podName);
-                            podsToRoll.put(podName, null);
+                            LOGGER.debugCr(reconciliation, "Pod {} does not need to be rolled", podContext.getPodName());
                         }
+                        clusterRollContext.add(podContext);
                     }
 
-                    if (needsRolling)   {
-                        return Future.succeededFuture(podsToRoll);
+                    if (clusterRollContext.requiresRestart())   {
+                        return Future.succeededFuture(clusterRollContext);
                     } else {
                         return Future.succeededFuture(null);
                     }
-                }).compose(podsToRoll -> {
-                    if (podsToRoll != null)  {
+                }).compose(clusterRollContext -> {
+                    if (clusterRollContext != null)  {
                         Promise<Void> promise = Promise.promise();
-                        Future<String> leaderFuture = leaderFinder.findZookeeperLeader(reconciliation, podsToRoll.keySet(), clusterCaSecret, coKeySecret);
+                        Future<String> leaderFuture = leaderFinder.findZookeeperLeader(reconciliation, clusterRollContext.podNames(), clusterCaSecret, coKeySecret);
 
                         leaderFuture.compose(leader -> {
                             LOGGER.debugCr(reconciliation, "Zookeeper leader is " + (ZookeeperLeaderFinder.UNKNOWN_LEADER.equals(leader) ? "unknown" : "pod " + leader));
                             Future<Void> fut = Future.succeededFuture();
 
                             // Then roll each non-leader pod => the leader is rolled last
-                            for (Map.Entry<String, List<String>> podEntry : podsToRoll.entrySet())  {
-                                if (podEntry.getValue() != null && !podEntry.getValue().isEmpty()) {
-                                    if (!podEntry.getKey().equals(leader)) {
-                                        LOGGER.debugCr(reconciliation, "Restarting non-leader pod {}", podEntry.getKey());
-                                        // roll the pod and wait until it is ready
-                                        // this prevents rolling into faulty state (note: this applies just for ZK pods)
-                                        fut = fut.compose(ignore -> restartPod(reconciliation, podEntry.getKey(), podEntry.getValue()));
-                                    } else {
-                                        LOGGER.debugCr(reconciliation, "Deferring restart of leader {}", podEntry.getKey());
+                            for (ZookeeperPodContext podContext : clusterRollContext.getPodContextsWithNonReadyFirst())  {
+                                final String podName = podContext.getPodName();
```

Comment:
Why do we need this variable? Its just a getter which should be easy to call directly, or?

Reply:
It had 6 usages in the block, so it was a reflex to pull it out. Happy to inline.

Reply:
It does not do any complex computation. So I would just call the method inline.